Allow whitespaces in search of invoice references

guewen opened the pull request:
When searching in Vendor references, we want to ignore the whitespaces.
This is because very often, it will contain BVR/ESR number that contains
whitespaces. And the user usually searches for the invoice number which
is only a part of the BVR/ESR number.

Example:
* the vendor reference on the supplier invoice is: 27 29990 00000 00001 70400 25019
* the user knows the invoice's supplier reference: 17040025
* when searching 17040025, he expects to find the invoice
* the reverse situation, where the ref contains no whitespaces and the
  user searches for a string with whitespaces might also happen

Commit:
Allow whitespaces in search of invoice references

When searching in Vendor references, we want to ignore the whitespaces.
This is because very often, it will contain BVR/ESR number that contains
whitespaces. And the user usually searches for the invoice number which
is only a part of the BVR/ESR number.

Example:
* the vendor reference on the supplier invoice is: 27 29990 00000 00001 70400 25019
* the user knows the invoice's supplier reference: 17040025
* when searching 17040025, he expects to find the invoice
* the reverse situation, where the ref contains no whitespaces and the
  user searches for a string with whitespaces might also happen

yvaucher reviewed (approved):
Nice change

gurneyalex commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -11,6 +11,37 @@ class AccountInvoice(models.Model):
+                     (operator,))
+            self.env.cr.execute(query, (value,))
+            ids = [t[0] for t in self.env.cr.fetchall()]
+            domain.append(('id', 'in', ids))
```
Don't you need to special if `ids == []`?

guewen replied:
if `ids == []`, the ORM automatically translates it to:

```
SELECT "account_invoice".id FROM "account_invoice" WHERE FALSE ORDER BY "account_invoice"."date_invoice" DESC,"account_invoice"."number" DESC,"account_invoice"."id" DESC
```

gurneyalex commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -11,6 +11,37 @@ class AccountInvoice(models.Model):
+                domain.append(arg)
+                continue
+            if value:
+                value = value.replace(' ', '')
```
I think you need to handle the special case where value.replace(' ', '') -> empty string. This will match *all* invoices and kill your perfs, and it may happen with incremental searches.

gurneyalex reviewed (commented):
a few questions asked.

Commit:
Remove support of strict operators

Only the like (&co.) operators will do the fuzzy search without
whitespaces. Strict equality operators will remain strict, which means
they will need to have an exact match. This what we would the most
expect from '='. Users will use the like/ilike from the UI and that's
only where whe should be tolerant regarding extra or missing whitespaces.

It fixes along the way:

* =like and =ilike where not working as they are not SQL operators
* when the original value only contains spaces, it would generate a
  costly query that matches all the records, now there is a fast path

guewen commented:
Added a new commit that remove the path for the operators which are not in the LIKE family, they will now be treated as usual. Only the LIKE (and family) operators will now search without whitespaces. The commit includes some details.
